hera_sim/defaults.py

```diff
@@ -0,0 +1,185 @@
+                  }
+
+class Defaults:
+    """
+    This class handles the retreival of simulation default parameters from
+    YAML files and the ability to switch the default settings while in an
+    interactive environment.
+    """
+
+    def __init__(self, config_file='h1c'):
+        self._config = self._get_config(config_file)
+        self._check_config()
+        self._use_season_defaults = False
+
+    """
+    Instantiate a Defaults object with a hook to a configuration file.
+
+    Args:
+        config_file (str, optional): hook to a YAML file or a season name
+            The currently supported observing seasons are 'h1c' and 'h2c'.
+            The loaded YAML file is intended to have the following form:
+            {module: {model: {param: default}}}, where 'module' is any one
+            of the modules
+    """
+
+    def __call__(self, module, model):
+        with open(self._config, 'r') as config:
+            defaults = yaml.load(config.read(), Loader=yaml.FullLoader)[module][model]
+        # handle instances where default parameters are related to interpolators
+        return self._retrieve_models(defaults)
+
+    def set_defaults(self, new_config):
+        self._config = self._get_config(new_config)
+        self._check_config()
+
+    def activate_defaults(self):
+        self._use_season_defaults = True
+
+    def deactivate_defaults(self):
+        self._use_season_defaults = False
+
+    def _get_config(self, config_file):
+        assert isinstance(config_file, str), \
+                "Default configurations are set by passing a hook to a " \
+                "configuration file or one of the season keys. The " \
+                "currently supported season configurations are " \
+                "{}.".format(SEASON_CONFIGS.keys())
+
+        if config_file in SEASON_CONFIGS.keys():
+            return SEASON_CONFIGS[config_file]
+        else:
+            return config_file
+
+    def _check_config(self):
+        assert path.exists(self._config), \
+                "Please ensure that a path to the specified configuration " \
+                "file exists. {} could not be found".format(self._config)
+
+    @property
+    def _version_is_compatible(self):
+        version = sys.version_info
+        if version.major < 3 or version.major > 3 and version.minor < 4:
+            warnings.warn("You are using a version of Python that is not " \
+                          "compatible with the Defaults class. If you would " \
+                          "like to use the features of the Defaults class, " \
+                          "then please use a version of Python newer than 3.4.")
+            return False
+        else:
+            return True
+
+    def _handler(self, func, *args, **kwargs):
+        if self._version_is_compatible:
+            @functools.wraps(func)
+            def new_func(*args, **kwargs):
+            # XXX do we want to use the season defaults by default?
```

Comment:
Thinking about this a bit more, perhaps it would be good to have a standalone function, say `set_defaults(config)`, which the user can call. Internally, it just does `_defaults._get_config(config)` then `activate_defaults`. I think whenever the user actually manually _sets_ defaults, the behaviour should be to use them everywhere, rather than func sig.

Reply:
There already is a `set_defaults` method, which changes the location of the configuration file pointed to. If `Defaults._use_season_defaults` is `True` (which is what happens when `Defaults.activate_defaults()` is called), then these new defaults will be used everywhere.

I'll update the initial comment in this PR to outline the basic functionality of the class and new additions so that it's very clear what the class actually does.

Reply:
Yeah, that became more clear as I read through the code, but it would be good to front-load the example usage :-).

I still stand behind the idea that _explicitly_ calling `set_defaults` (or `set()`, as I would have it) should by default also set `activate_defaults` (or `activate()` as I would have it).

Also, just thinking, is `Defaults` meant to be a singleton? I.e. is `defaults` meant to be the only ever instance of `Defaults`? The user is not supposed to make their own instance, right, just to use the pre-made one? In that case, we should hide `Defaults --> _Defaults`.

Reply:
Yes, `Defaults` is meant to be a singleton. I suppose we could hide it, but the package `__init__` method only loads in the `defaults` instance of `Defaults` from the `defaults` module (sorry if that sounds confusing), so that the `defaults` module is not discoverable and an end-user should not be able to make their own `Defaults` instance.

Reply:
I would still hide it with an underscore, as that is the convention.